modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotFutureTask.java

```diff
@@ -0,0 +1,882 @@
+                        new PageStoreSerialWriter(store,
+                            partDeltaFile(cacheWorkDir(tmpConsIdDir, cacheDirName(gctx.config())), partId)));
+
+                    partFileLengths.put(pair, store.size());
+                }
+            }
+        }
+        catch (IgniteCheckedException e) {
+            acceptException(e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onCheckpointBegin(Context ctx) {
+        if (stopping())
+            return;
+
+        assert !processed.isEmpty() : "Partitions to process must be collected under checkpoint mark phase";
+
+        wrapExceptionIfStarted(() -> snpSndr.init(processed.values().stream().mapToInt(Set::size).sum()))
+            .run();
+
+        // Snapshot task can now be started since checkpoint write lock released and
+        // there is no error happen on task init.
+        if (!startedFut.onDone())
+            return;
+
+        // Submit all tasks for partitions and deltas processing.
+        List<CompletableFuture<Void>> futs = new ArrayList<>();
+
+        if (log.isInfoEnabled())
+            log.info("Submit partition processing tasks with partition allocated lengths: " + partFileLengths);
+
+        Collection<BinaryType> binTypesCopy = cctx.kernalContext()
+            .cacheObjects()
+            .metadata(Collections.emptyList())
+            .values();
+
+        // Process binary meta.
+        futs.add(CompletableFuture.runAsync(
+            wrapExceptionIfStarted(() -> snpSndr.sendBinaryMeta(binTypesCopy)),
+            snpSndr.executor()));
+
+        List<Map<Integer, MappedName>> mappingsCopy = cctx.kernalContext()
+            .marshallerContext()
+            .getCachedMappings();
+
+        // Process marshaller meta.
+        futs.add(CompletableFuture.runAsync(
+            wrapExceptionIfStarted(() -> snpSndr.sendMarshallerMeta(mappingsCopy)),
+            snpSndr.executor()));
+
+        FilePageStoreManager storeMgr = (FilePageStoreManager)cctx.pageStore();
+
+        for (Map.Entry<Integer, Set<Integer>> e : processed.entrySet()) {
+            int grpId = e.getKey();
+
+            CacheGroupContext gctx = cctx.cache().cacheGroup(grpId);
+
+            if (gctx == null) {
+                acceptException(new IgniteCheckedException("Cache group context has not found " +
+                    "due to the cache group is stopped: " + grpId));
+
+                break;
+            }
+
+            // Process the cache group configuration files.
+            futs.add(CompletableFuture.runAsync(
+                wrapExceptionIfStarted(() -> {
+                    List<File> ccfgs = storeMgr.configurationFiles(gctx.config());
+
+                    if (ccfgs == null)
+                        return;
+
+                    for (File ccfg0 : ccfgs)
+                        snpSndr.sendCacheConfig(ccfg0, cacheDirName(gctx.config()));
+                }),
+                snpSndr.executor())
```

Comment:
Looks like there is a race here: we read the configuration file from snapshot executor thread while the files can be modified by SQL create/drop column commands, for example.

Reply:
You right. I've completely missed that cache configuration files can be changed during SQL add/drop a column, create/drop an index.

I've fixed it and add an appropriate test for such a case.